src/utils/page-functions.js

```diff
@@ -0,0 +1,151 @@
+import path from 'path'
+import webpack from 'webpack'
+import {
+  readFile,
+  exists,
+  stats,
+  glob,
+  camelCase,
+  findNodeModulesPath,
+  createCacheKey,
+  getCachePath
+} from '.'
+
+export async function createPageFunctions(page, sourceFiles, babelPresets) {
+  const pageFunctions = {}
+
+  if (!sourceFiles) {
+    return pageFunctions
+  }
+
+  if (typeof sourceFiles === 'string') {
+    sourceFiles = await glob(sourceFiles)
+  }
+
+  for (let file of sourceFiles) {
+    let fnName
+    if (Array.isArray(file)) {
+      [file, fnName] = file
+    } else {
+      fnName = camelCase(path.basename(file, '.js'))
+    }
+
+    if (pageFunctions[fnName]) {
+      // eslint-disable-next-line no-console
+      console.warn(`A page function with name '${fnName}' already exists, the old one will be overwritten`)
+    }
+
+    pageFunctions[fnName] = async (...args) => {
+      const body = await getPageFunctionBody(fnName, file, babelPresets)
+
+      return page.runAsyncScript({
+        args: [],
+        body: `${body} return (pageFn.${fnName} || pageFn.default).apply(null, arguments)`
+      }, ...args)
+    }
+  }
+
+  return pageFunctions
+}
+
+export async function getPageFunctionBody(fnName, filePath, babelPresets) {
+  const cacheKey = createCacheKey(filePath, babelPresets)
+  const cacheFile = `${fnName}-${cacheKey}.js`
+  const cachePath = await getCachePath(cacheFile)
+  let cacheValid = await exists(cachePath)
+
+  if (cacheValid) {
```

Comment:
@pi0 Checking just the mtime of the source file is not sufficient when you use imports in that source file and have changed something in of those. Do you maybe have any recommendations?

Eg we could just mtime for any file in `path.dirname` of the source file but thats only a partial fix. Or maybe support a NODE_ENV = `test/dev` which you could set when you are writing your tests